internal/fnruntime/runner.go

```diff
@@ -66,27 +69,94 @@ type FunctionRunner struct {
 	name          string
 	disableOutput bool
 	filter        *runtimeutil.FunctionFilter
+	fnResults     *fnresult.ResultList
 }
 
 func (fr *FunctionRunner) Filter(input []*yaml.RNode) (output []*yaml.RNode, err error) {
+
 	if fr.disableOutput {
-		output, err = fr.filter.Filter(input)
+		output, err = fr.do(input)
 	} else {
 		pr := printer.FromContextOrDie(fr.ctx)
 		printOpt := printer.NewOpt()
 		pr.OptPrintf(printOpt, "[RUNNING] %q\n", fr.name)
-		output, err = fr.filter.Filter(input)
+		output, err = fr.do(input)
 		if err != nil {
 			pr.OptPrintf(printOpt, "[FAIL] %q\n", fr.name)
+			var fnErr *errors.FnExecError
+			if goerrors.As(err, &fnErr) {
+				pr.OptPrintf(printOpt.Stderr(), "%s\n", fnErr.String())
+				return nil, errors.ErrAlreadyHandled
+			}
 			return nil, err
 		}
 		// capture the result from running the function
 		pr.OptPrintf(printOpt, "[PASS] %q\n", fr.name)
 	}
+	return output, err
+}
+
+/*
+func printResults(ctx context.Context, fnResult *fnresult.Result) {
+	pr := printer.FromContextOrDie(ctx)
+	printOpt := printer.NewOpt()
+	for _, item := range fnResult.Results {
+		pr.OptPrintf(printOpt, "\t[%s] %s  : \n", strings.ToUpper(string(item.Severity)), item.Message)
+	}
+} */
+
+func (fr *FunctionRunner) do(input []*yaml.RNode) (output []*yaml.RNode, err error) {
+	fnResult := &fnresult.Result{Image: fr.name}
 
-	// TODO(droot): print functionResults
+	output, err = fr.filter.Filter(input)
+
+	// parse the results irrespective of the success/failure of fn exec
+	resultErr := parseStructuredResult(fr.filter.Results, fnResult)
+	if resultErr != nil {
+		// Not sure if it's a good idea. This may mask the original
+		// function exec error. Revisit this if this turns out to be true.
+		return output, resultErr
+	}
+	if err != nil {
+		var fnErr *errors.FnExecError
+		if goerrors.As(err, &fnErr) {
+			fnResult.ExitCode = fnErr.ExitCode
+			fnResult.Stderr = fnErr.Stderr
+			// fnErr.FnResult = fnResult
+			fr.fnResults.ExitCode = 1
+		}
+		// accumulate the results
+		fr.fnResults.Items = append(fr.fnResults.Items, *fnResult)
+		return output, err
+	}
+	fnResult.ExitCode = 0
+	fr.fnResults.Items = append(fr.fnResults.Items, *fnResult)
+	return output, nil
+}
 
-	return
+func parseStructuredResult(yml *yaml.RNode, fnResult *fnresult.Result) error {
+	if yml.IsNilOrEmpty() {
+		return nil
+	}
+	// TODO(droot): checking for legacy structured result format
```

Comment:
what's legacy here? AFAIK, the wire format of results in Function Spec Standard is not changing.

Reply:
Actually, we are changing the wire format. 

`kubeeval:v0.1` implements new format where result items are not wrapped

```
- message: selector is required
  severity: error
  resourceRef:
    apiVersion: apps/v1
    kind: Deployment
    namespace: ''
    name: nginx-deployment
  file:
    path: resources.yaml
  field:
    path: selector
- message: template is required
  severity: error
  resourceRef:
    apiVersion: apps/v1
    kind: Deployment
    namespace: ''
    name: nginx-deployment
  file:
    path: resources.yaml
  field:
    path: template
```
Old format: `search-replace:v0.1` is using the old format where result items are wrapped.

```
name: search-replace
items:
  - message: Mutated field value to "4"
    field:
      path: spec.replicas
    file:
      path: resources.yaml
```

While implementing this I realized the older format has more flexibility because it has a container where items are stored, so if we want to introduce any new field in the future or even change the content of the result items or support new type of formats.

For now, I am supporting both.

@Shell32-Natsu @mengqiy

Reply:
so @mengqiy confirmed that `kubeval` is using the old format. `search-replace` is using new format. May be @phanimarupaka can share some info.

Reply:
@droot I am using kyaml v0.10.10(https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/master/functions/go/search-replace/go.mod#L8). Is this considered new or old? The results are wrapped, but the latest version of kyaml has different Result struct.

Reply:
General question: Should a function use same version of kyaml as kpt ? Looks like it should, else the result is broken runtime.

Reply:
I think we should support both here.

Reply:
So GO SDK uses a result format https://pkg.go.dev/sigs.k8s.io/kustomize/kyaml@v0.10.19/fn/framework#Result that is a struct (MapNode in yaml parsing context) and it contains `items` field that contains result items.
```Go
type Result struct {
	// Name is the name of the function creating the result
	Name string `yaml:"name,omitempty"`

	// Items are the individual results
	Items []ResultItem `yaml:"items,omitempty"`
}
```
while Typescript SDK uses https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/b7e346f0c18f972edf7ca02272734fbc68fca905/ts/kpt-functions/src/types.ts#L306 list of result items as `.Result` field so it's a sequence node (in context of yaml parsing).

So our plan is:
1. To support both the formats for now.
2. Update the function spec to have one canonical format (Note: this is not documented and that resulted in this confusion)
3. Update the SDKs to use same function spec